src/plan/generational/immix/gc_work.rs

```diff
@@ -0,0 +1,186 @@
+        bytes: usize,
+        align: usize,
+        offset: isize,
+        _semantics: crate::AllocationSemantics,
+    ) -> Address {
+        debug_assert!(VM::VMActivePlan::global().base().gc_in_progress_proper());
+        if self.plan.immix.in_defrag() {
+            self.defrag_copy.alloc(bytes, align, offset)
+        } else {
+            self.copy.alloc(bytes, align, offset)
+        }
+    }
+
+    #[inline(always)]
+    fn post_copy(
+        &mut self,
+        obj: ObjectReference,
+        tib: Address,
+        bytes: usize,
+        semantics: crate::AllocationSemantics,
+    ) {
+        // Missing ImmixSpace.post_copy()
+        crate::plan::generational::generational_post_copy::<VM>(obj, tib, bytes, semantics)
+    }
+}
+
+impl<VM: VMBinding> GenImmixCopyContext<VM> {
+    pub fn new(mmtk: &'static MMTK<VM>) -> Self {
+        let plan = &mmtk.plan.downcast_ref::<GenImmix<VM>>().unwrap();
+        Self {
+            plan,
+            // it doesn't matter which space we bind with the copy allocator. We will rebind to a proper space in prepare().
+            copy: ImmixAllocator::new(
+                VMThread::UNINITIALIZED,
+                Some(&plan.immix),
+                &*mmtk.plan,
+                false,
+            ),
+            defrag_copy: ImmixAllocator::new(
+                VMThread::UNINITIALIZED,
+                Some(&plan.immix),
+                &*mmtk.plan,
+                true,
+            ),
+        }
+    }
+}
+
+impl<VM: VMBinding> GCWorkerLocal for GenImmixCopyContext<VM> {
+    fn init(&mut self, tls: VMWorkerThread) {
+        CopyContext::init(self, tls);
+    }
+}
+
+use crate::plan::immix::gc_work::TraceKind;
+
+/// ProcessEdges for a full heap GC for generational immix. The const type parameter
+/// defines whether there is copying in the GC.
+/// Note that even with TraceKind::Fast, there is no defragmentation, we are still
+/// copying from nursery to immix space. So we always need to write new object
+/// references in process_edge() (i.e. we do not need to overwrite the default implementation
+/// of process_edge() as the immix plan does).
+pub(super) struct GenImmixMatureProcessEdges<VM: VMBinding, const KIND: TraceKind> {
+    plan: &'static GenImmix<VM>,
+    base: ProcessEdgesBase<GenImmixMatureProcessEdges<VM, KIND>>,
+    mmtk: &'static MMTK<VM>,
+}
+
+impl<VM: VMBinding, const KIND: TraceKind> ProcessEdgesWork
+    for GenImmixMatureProcessEdges<VM, KIND>
+{
+    type VM = VM;
+
+    fn new(edges: Vec<Address>, _roots: bool, mmtk: &'static MMTK<VM>) -> Self {
+        let base = ProcessEdgesBase::new(edges, mmtk);
+        let plan = base.plan().downcast_ref::<GenImmix<VM>>().unwrap();
+        Self { plan, base, mmtk }
+    }
+
+    #[cold]
+    fn flush(&mut self) {
+        use crate::policy::immix::ScanObjectsAndMarkLines;
+        use crate::scheduler::WorkBucketStage;
+        let mut new_nodes = vec![];
+        std::mem::swap(&mut new_nodes, &mut self.nodes);
```

Comment:
The implementation of `GenImmixMatureProcessEdges.flush()` is simply duplicating `ImmixProcessEdges.flush()`. Also there is no early return for the default implementation of `ProcessEdgesWork.flush()`.
https://github.com/mmtk/mmtk-core/blob/198a142683fd970d4fcf495affb5f5374ea39e04/src/plan/immix/gc_work.rs#L135-L145

I don't think we should have `self.nodes` to be empty. It should be checked here:
https://github.com/mmtk/mmtk-core/blob/198a142683fd970d4fcf495affb5f5374ea39e04/src/scheduler/gc_work.rs#L471-L473

I will add an assertion to make sure `self.nodes` is not empty in those 3 places.

Reply:
I have done some refactoring in https://github.com/mmtk/mmtk-core/pull/449/commits/469454b1c3cb5330ae2c661d5d54f75029b3c4fd.

Reply:
Actually the assertion may fail, because we also call `flush()` in the finalizer processor, which does not check if nodes are empty or not. I have added an early return for `flush()`.